apps/base-docs/src/components/DocChat/docChat.ts

```diff
@@ -0,0 +1,257 @@
+  setIsLoading: (isLoading: boolean) => void,
+  isGenerating: boolean,
+  setIsGenerating: (isGenerating: boolean) => void,
+  chatHistory: ChatHistoryMessage[],
+  setChatHistory: (chatHistory: (prevState: ChatHistoryMessage[]) => ChatHistoryMessage[]) => void,
+  setConversation: (
+    conversation: (prevState: ConversationMessage[]) => ConversationMessage[],
+  ) => void,
+) {
+  try {
+    let fullResponse = '';
+    let responseSources: ResponseSource[];
+    let messageId: number;
+
+    controller = new AbortController();
+
+    logGptEvent('gpt_prompt_submitted', {
+      conversation_id: conversationId,
+      prompt,
+    });
+
+    const url = 'https://api.mendable.ai/v1/mendableChat';
+
+    const data = {
+      api_key: '7f73a4b9-96b3-4689-8688-b120ea0ee916',
+      question: prompt,
+      history: chatHistory,
+      conversation_id: conversationId,
+      retriever_options: {
+        num_chunks: 4,
+      },
+    };
+
+    await fetchEventSource(url, {
+      method: 'POST',
+      headers: {
+        Accept: 'text/event-stream',
+        'Content-Type': 'application/json',
+      },
+      openWhenHidden: true,
+      body: JSON.stringify(data),
+      signal: controller.signal,
+      onmessage(event: unknown) {
+        const parsedData = JSON.parse(event.data);
+        const chunk: string = parsedData.chunk;
+
+        if (chunk === '<|source|>') {
+          responseSources = parsedData.metadata as ResponseSource[];
+          return;
+        } else if (chunk === '<|message_id|>') {
+          messageId = parsedData.metadata as number;
+          return;
+        }
+
+        // End loading spinner and show Stop Generating button
+        if (!isGenerating) {
+          setIsLoading(false);
+          setIsGenerating(true);
+        }
+
+        // Update full response string
+        fullResponse = fullResponse.concat(chunk);
+
+        // Update rendered conversation data for current response
+        setConversation((prevState: ConversationMessage[]) => {
+          const currentResponse = prevState.slice(-1)[0];
+          const updatedResponse = {
+            ...currentResponse,
+            content: currentResponse.content.concat(chunk),
+          };
+
+          const newState = [...prevState.slice(0, -1), updatedResponse];
+          return newState;
+        });
+
+        return;
+      },
+      onclose() {
+        // Add Mendable message ID and sources to current response data
+        const sourceURLs: string[] = responseSources.map((source) => source.link);
+
+        setConversation((prevState: ConversationMessage[]) => {
+          const currentResponse = prevState.slice(-1)[0];
+          const updatedResponse = {
+            ...currentResponse,
+            sources: sourceURLs,
+            helpful: null,
+            messageId,
+          };
+
+          const newState = [...prevState.slice(0, -1), updatedResponse];
+          setSessionConversation(newState);
+
+          return newState;
```

Comment:
Ditto above

Reply:
This is the only state update I didn't simplify since `newState` needs to be passed to `setSessionConversation` before it's returned for the `setConversation` update.